macaroonbakery/checker.py

```diff
@@ -36,4 +47,391 @@ def __hash__(self):
+# LOGIN_OP represents a login (authentication) operation.
+# A macaroon that is associated with this operation generally
+# carries authentication information with it.
+LOGIN_OP = Op(entity='login', action='login')
+
+
+class Checker(object):
+    '''Checker implements an authentication and authorization checker.
+
+    It uses macaroons as authorization tokens but it is not itself responsible
+    for creating the macaroons
+    See the Oven type (TODO) for one way of doing that.
+    '''
+    def __init__(self, checker=checkers.Checker(),
+                 authorizer=ClosedAuthorizer(),
+                 identity_client=NoIdentities(),
+                 macaroon_opstore=None):
+        '''
+        :param checker: a first party checker implementing a
+        FirstPartyCaveatChecker.
+        :param authorizer (Authorizer): used to check whether an authenticated
+        user is allowed to perform operations.
+        The identity parameter passed to authorizer.allow will always have been
+        obtained from a call to identity_client.declared_identity.
+        :param identity_client (IdentityClient) used for interactions with the
+        external identity service used for authentication.
+        If this is None, no authentication will be possible.
+        :param macaroon_opstore (object with new_macaroon and macaroon_ops
+        method): used to retrieve macaroon root keys and other associated
+        information.
+        '''
+        self._first_party_caveat_checker = checker
+        self._authorizer = authorizer
+        self._identity_client = identity_client
+        self._macaroon_opstore = macaroon_opstore
+
+    def auth(self, *mss):
+        ''' Returns a new AuthChecker instance using the given macaroons to
+        inform authorization decisions.
+        '''
+        return AuthChecker(parent=self,
+                           macaroons=mss)
+
+    def namespace(self):
+        ''' Returns the namespace of the first party checker.
+        '''
+        return self._first_party_caveat_checker.namespace()
+
+
+class AuthChecker(object):
+    '''Authorizes operations with respect to a user's request.
+
+    The identity is authenticated only once, the first time any method
+    of the AuthChecker is called, using the context passed in then.
+
+    To find out any declared identity without requiring a login,
+    use allow(ctx); to require authentication but no additional operations,
+    use allow(ctx, LOGIN_OP).
+    '''
+    mutex = Lock()
+
+    def __init__(self, parent, macaroons):
+        '''
+
+        :param parent (Checker): used to check first party caveats.
+        :param macaroons: a list of py macaroons
+        '''
+        self._macaroons = macaroons
+        self._init_errors = []
+        self._executed = False
+        self._identity = None
+        self._identity_caveats = []
+        self.parent = parent
+        self._conditions = None
+
+    def _init(self, ctx):
+        try:
```

Comment:
You don't need to call acquire and release manually:

    with self.mutex:
        if not etc

should work fine (assuming the lock is in the instance not the class).